### PR TITLE
Feature/subject restriction endpoint

### LINKED ISSUE #300: Create API endpoint in the back-end for the right to restriction
Create an endpoint that allows a subject to restrict the use of his data.
This will require updates in the subjects/subjects routes, controller and service. 
It will also require updates to the subjects_processor table, and hence migration, in order to contain what each processor is allowed to do with the subject's data.
This endpoint should call the smart contract function recordRestrictionByController and also update the DB subjects_processors table.

### LINKED ISSUE #311: Call recordAccessByController in the access-data endpoint
That endpoint is missing the function that acts inside the smart contract.
It turns out it's not going to be so simple because the processors code is reaching into the subjects domain.
Blocked by #329

### PR DESCRIPTION
**Pull request by sindelio**
Closes #300

Create subject/restrict and subject/get-restrictions endpoints, and add a validator for the first.
Tested in Postman, test suite not yet ready.

### CODE REVIEW

File: cg/migrations/20180402122250_subjects_and_keys.js [hunk 1]
```diff
@@ -7,6 +7,9 @@ exports.up = function(knex, Promise) {
         .primary();
       table.timestamps(false, true);
       table.text('personal_data').notNullable();
+      table.boolean('direct_marketing').notNullable();
```
[the remark this answers is not in the document]
Comment (sindelio) [reply cg/migrations/20180402122250_subjects_and_keys.js]: Oh I see. This will require everyone to run the setup again, after resetting the volumes

[the remark this answers is not in the document]
Comment (Jakedavies) [reply cg/migrations/20180402122250_subjects_and_keys.js]: no all that will need to be done is `docker-compose exec admin yarn run knex migrate:latest`

File: cg/src/domains/subjects/subjects.service.js [hunk 2]
```diff
@@ -207,14 +212,15 @@ class SubjectsService {
...
+    const [ data ] = await this.db('subjects')
       .join('subject_keys', 'subjects.id', '=', 'subject_keys.subject_id')
       .select('personal_data')
       .select('key')
       .where({ subject_id: subjectId });
+      
     if (!data) throw new NotFound('Subject not found');
     const decryptedData = decryptFromStorage(data.personal_data, data.key);
-    //There should be a smart contract call here!
+    //await recordAccessByController(subjectId); //Fixing this won't be so simple
```
Comment (Jakedavies) [inline cg/src/domains/subjects/subjects.service.js]: ?

File: cg/test/processors/processors.api.spec.js [hunk 3]
```diff
@@ -58,14 +58,17 @@ describe('Processor requesting data', () => {
   });
 
   it('should respond with the decrypted data for the subject requested', async () => {
+    // Given
     const key = encryption.generateClientKey();
     const testData = { test: 'true' };
     const subjectId = hash('p1');
     await db('subjects').insert({
```
Comment (Jakedavies) [inline cg/test/processors/processors.api.spec.js]: it would be valuable to create a helper that sets up the subject for you at this point, lots of setup code could be removed.

File: frontend/src/containers/Share.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+// CG Dependencies
+import Element from '../elements/components/Element';
+import Elements from '../elements/components/Elements';
+Elements();
+
+class Share extends React.Component {
+  render() {
+    const ShareData = Element.create({
+      source: 'share-data'
+    });
```
Comment (Jakedavies) [inline frontend/src/containers/Share.js]: why is this here?

File: cg/src/domains/subjects/subjects.routes.js [hunk 5]
```diff
@@ -83,10 +84,26 @@ module.exports = app => {
 
   router.post(
     '/initiate-rectification',
-    rectificationRequestValidator,
+    rectificationValidator,
     asyncHandler(async (req, res) => subjectsController.initiateRectification(req, res))
   );
 
+  //Probably more endpoints should be controllerOnly
+  //Also, it would be good to run a check on the interaction between the rights, such as erasing a subject and trying stuff when his data has been erased
```
[the remark this answers is not in the document]
Comment (sindelio) [reply cg/src/domains/subjects/subjects.routes.js]: Ok, was a reminder

File: cg/src/domains/subjects/subjects.service.js [hunk 6]
```diff
@@ -236,6 +242,39 @@ class SubjectsService {
     });
     return { success: true };
   }
+
+  async restrict(subjectId, directMarketing, emailCommunication, research){
```
[the remark this answers is not in the document]
Comment (sindelio) [reply cg/src/domains/subjects/subjects.service.js]: No xD, will do it

File: cg/src/domains/subjects/subjects.service.js [hunk 7]
```diff
@@ -236,6 +242,39 @@ class SubjectsService {
...
+        direct_marketing: directMarketing,
+        email_communication: emailCommunication,
+        research: research,
+      })
+
+    await recordRestrictionByController(subjectId, directMarketing, emailCommunication, research);    
+  }
+
+  async getRestrictions(subjectId){
+    const [ subjectExists ] = await this.db('subjects')
```
[the remark this answers is not in the document]
Comment (sindelio) [reply cg/src/domains/subjects/subjects.service.js]: I was going to ask about that. Fixing it

File: cg/src/domains/subjects/subjects.service.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (sindelio) [reply cg/src/domains/subjects/subjects.service.js]: Done

**Comment by sindelio** [comment]
It's because the PR is based on my last PR, which is not updated. After I address the changes, the PR will be reopened based on master
